pkg/client/client.go

```diff
@@ -215,6 +214,7 @@ func (c *client) startCheckin() {
 			checkinCtx, checkinCancel := context.WithCancel(c.ctx)
```

Comment:
it is safe to call cancel multiple times. But never forget to call cancel => use defer always. If you are inside a loop consider to factor out the body into a function. 

The function body is huge with many go-routines. Consider to split it up even further.